pybamm/simulation.py

```diff
@@ -454,6 +448,20 @@ def solve(
                 pybamm.logger.info(self.experiment.operating_conditions_strings[idx])
                 inputs.update(exp_inputs)
                 kwargs["inputs"] = inputs
+                # If the last time entry was due to an event take a single step
+                # to get the solution back on to the correct reporting period.
+                # e.g. if the period is 30s and an event is triggered at 32s we
+                # return the times [0, 30, 32, 60] instead of [0, 30, 32, 62]
+                if hasattr(self._solution, "t_event"):
+                    if self._solution.t_event is not None and (
+                        self._solution.t_event[0] == self.solution.t[-1]
+                    ):
+                        dt_event = (
+                            exp_inputs["period"]
+                            - (self._solution.t[-1] - self._solution.t[-2])
+                            * self._solution.timescale_eval
+                        )
```

Comment:
`dt_event` can be negative here if switching from a longer period (e.g. 180s) to a shorter period (e.g. w60s), since in that case `(self._solution.t[-1] - self._solution.t[-2]) * self._solution.timescale_eval` (e.g. 150s) can be longer than the new period even though it is shorter than the old period